File: core/src/main/java/org/opencds/cqf/cql/ls/core/utility/Converters.java
package org.opencds.cqf.cql.ls.core.utility;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import kotlinx.io.Buffer;
import kotlinx.io.Source;

public class Converters {

    public static String inputStreamToString(InputStream inputStream) throws IOException {
        StringBuilder resultStringBuilder = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (!resultStringBuilder.isEmpty()) resultStringBuilder.append("\n"); // Append newline if needed
                resultStringBuilder.append(line);
            }
        }
        return resultStringBuilder.toString();
    }

    public static Source stringToSource(String text) {
        Buffer buffer = new Buffer();
        // Write the string to the buffer using a specific character encoding
        buffer.write(text.getBytes(), 0, text.length());
        // Return the buffer as a Source
        return buffer;
    }

    public static Source inputStreamToSource(InputStream inputStream) throws IOException {
        return stringToSource(inputStreamToString(inputStream));
    }
}

File: core/src/main/java/org/opencds/cqf/cql/ls/core/utility/Converters.kt
package org.opencds.cqf.cql.ls.core.utility

import kotlinx.io.Buffer
import kotlinx.io.Source
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

object Converters {
    @JvmStatic
    @Throws(IOException::class)
    fun inputStreamToString(inputStream: InputStream): String {
        val sb = StringBuilder()
        BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8)).use { reader ->
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                if (sb.isNotEmpty()) sb.append("\n")
                sb.append(line)
            }
        }
        return sb.toString()
    }

    @JvmStatic
    fun stringToSource(text: String): Source {
        val buffer = Buffer()
        val bytes = text.toByteArray()
        buffer.write(bytes, 0, bytes.size)
        return buffer
    }

    @JvmStatic
    @Throws(IOException::class)
    fun inputStreamToSource(inputStream: InputStream): Source {
        return stringToSource(inputStreamToString(inputStream))
    }
}
